src/sinks/util/mod.rs

```diff
@@ -314,3 +320,93 @@ mod test {
         assert_eq!(15, ack_counter.load(Ordering::Relaxed));
     }
 }
+
+/// Tower Request based configuration
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
+pub struct TowerRequestConfig {
+    pub request_in_flight_limit: Option<usize>,        // 5
+    pub request_timeout_secs: Option<u64>,             // 60
+    pub request_rate_limit_duration_secs: Option<u64>, // 1
+    pub request_rate_limit_num: Option<u64>,           // 5
+    pub request_retry_attempts: Option<usize>,         // max_value()
+    pub request_retry_backoff_secs: Option<u64>,       // 1
+}
+
+impl TowerRequestConfig {
+    pub fn unwrap_with(&self, defaults: &TowerRequestConfig) -> TowerRequestSettings {
+        TowerRequestSettings {
+            in_flight_limit: self
+                .request_in_flight_limit
+                .or(defaults.request_in_flight_limit)
+                .unwrap_or(5),
+            timeout: Duration::from_secs(
+                self.request_timeout_secs
+                    .or(defaults.request_timeout_secs)
+                    .unwrap_or(60),
+            ),
+            rate_limit_duration: Duration::from_secs(
+                self.request_rate_limit_duration_secs
+                    .or(defaults.request_rate_limit_duration_secs)
+                    .unwrap_or(1),
+            ),
+            rate_limit_num: self
+                .request_rate_limit_num
+                .or(defaults.request_rate_limit_num)
+                .unwrap_or(5),
+            retry_attempts: self
+                .request_retry_attempts
+                .or(defaults.request_retry_attempts)
+                .unwrap_or(usize::max_value()),
+            retry_backoff: Duration::from_secs(
+                self.request_retry_backoff_secs
+                    .or(defaults.request_retry_backoff_secs)
+                    .unwrap_or(1),
+            ),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct TowerRequestSettings {
+    pub in_flight_limit: usize,
+    pub timeout: Duration,
+    pub rate_limit_duration: Duration,
+    pub rate_limit_num: u64,
+    pub retry_attempts: usize,
+    pub retry_backoff: Duration,
+}
+
+impl TowerRequestSettings {
+    pub fn retry_policy<L: RetryLogic>(&self, logic: L) -> FixedRetryPolicy<L> {
+        FixedRetryPolicy::new(self.retry_attempts, self.retry_backoff, logic)
+    }
+
+    pub fn batch_sink<B, L, S, T>(
+        &self,
+        retry_logic: L,
+        service: S,
+        acker: Acker,
+    ) -> BatchServiceSink<T, ConcurrencyLimit<RateLimit<Retry<FixedRetryPolicy<L>, Timeout<S>>>>, B>
```

Comment:
To me the return type is fine, at least I don't see a way to simplify it.

Reply:
Type alias can help here

Reply:
Do you mean an alias for the full `BatchServiceSink<...>` to reduce line length or some other part?

Reply:
Maybe something like this

```rust
type FixedTimeoutRetry<L, S> = Retry<FixedRetryPolicy<L>, Timeout<S>>;
type ConcurencyLimitFixedRetry<L, S> = ConcurrencyLimit<RateLimit<L, S>;
```

just to reduce number of repeating `> ` in the same expression?